src/metabase/models/humanization.clj

```diff
@@ -16,75 +19,111 @@
-    (db/update! model id
-      :display_name new-display-name)))
+  ;; Try all the different implementations of name->human-readable-name. If any (f internal-name) is equal to
+  ;; display-name we can safely assume that display-name was set automatically using that strategy and stop there.
+  ;; Otherwise if none of the strategies match (i.e. `some` returns `nil`) we can assume a custom name was set.
+  (nil? (some #(= display-name (% internal-name))
+              (vals (methods name->human-readable-name)))))
+
+(defn- re-humanize-names!
+  "Update all non-custom display names of all instances of `model` (e.g. Table or Field). To prevent explosions on very
+  large instances this is done in chunks of 10000 IDs at a time"
+  ;; TODO - once Toucan has the new streaming SELECT functions we can use those instead of manually chunking
+  ;; TODO - Alternatively, just make this chunking pattern generic and add to Toucan. Could be useful elsewhere!
+  ([model] (re-humanize-names! model 0))
+  ([model starting-id]
+   (let [chunk-size 10000]
+     (when-let [instances (seq (db/select [model :id :name :display_name]
+                                 :id [:>= starting-id]
+                                 {:limit chunk-size}))]
+       (doseq [{id :id, internal-name :name, display-name :display_name} instances
+               :let  [new-display-name (name->human-readable-name internal-name)]
+               :when (and (not= display-name new-display-name)
+                          (not (custom-display-name? internal-name display-name)))]
+         (log/info (format "Updating display name for %s '%s': '%s' -> '%s'"
+                           (name model) internal-name display-name new-display-name))
+         (db/update! model id
+           :display_name new-display-name))
+       ;; if we got a full chunk, do the next chunk, starting at one past the ID of the last object
+       (when (>= (count instances) chunk-size)
+         (recur model (inc (:id (last instances)))))))))
 
 (defn- re-humanize-table-and-field-names!
-  "Update the display names of all tables in the database using new values obtained from the (obstensibly toggled implementation of) `name->human-readable-name`."
+  "Update the non-custom display names of all Tables & Fields in the database using new values obtained from
+  the (obstensibly swapped implementation of) `name->human-readable-name`."
   []
   (re-humanize-names! 'Table)
   (re-humanize-names! 'Field))
 
 
-(defn- set-enable-advanced-humanization! [^Boolean new-value]
-  (setting/set-boolean! :enable-advanced-humanization new-value)
-  (log/info (format "Now using %s table name humanization." (if (enable-advanced-humanization) "ADVANCED" "SIMPLE")))
+(defn- set-humanization-strategy! [new-strategy]
+  ;; check to make sure `new-strategy` is a valid strategy, or throw an Exception it is it not.
+  (when-not (get-method name->human-readable-name (keyword new-strategy))
```

Comment:
Is the thinking here that the user would write an implementation of the name->human-readable-name multimethod and put it on the classpath, thus the need to ask the multimethod what implementation is has?

Reply:
Yes, or if we decided to add a new implementation ourselves my thinking was if we do things this way there are less things that need to be changed. Thus less opportunities to forget to do something.